### PR TITLE
Bug 1405410 — Rendering graph (including legend) from a test.

### PR DESCRIPTION
**Pull request by jhugman**
This PR introduces a GraphRepresentation via [graphviz][1] and utilities to write it to disk.

There is also a legend. This is made by using the same primitives used to render the graph, so should change as you style them. No attempt has been made to parameterize the graph representation so users can change the appearance. This should be considered once we get a designer to see how ugly the graph is.

We do not yet call graphviz from the test.

The graph file is dumped in `$HOME/Library/Caches/tools.mappamundi/graph.dot`, or `$HOME/Library/Caches/graph.dot` if you don't create that directory.

You can render the graph using:

```sh
% dot -Tpng $HOME/Library/Caches/tools.mappamundi/graph.dot -ograph.png
```

graph

The above is taken from rendering the graph of the test application.

https://bugzilla.mozilla.org/show_bug.cgi?id=1405410

 [1]: https://graphviz.gitlab.io/

### CODE REVIEW

File: Sources/GraphRepresentation.swift [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+            } else if let _ = node as? MMScreenActionNode {
+                renderer.renderScreenActionNode(name: name)
+            }
+        }
+
+        namedScenes.forEach { (src, node) in
+            if let node = node as? MMScreenStateNode {
+                node.edges.values.forEach { edge in
+                    guard let dest = namedScenes[edge.destinationName] else { return }
+                    let directed: Bool
```
Comment (justindarc) [inline Sources/GraphRepresentation.swift]: shouldn't this be declared inside the `if` below?

File: Sources/GraphRepresentation.swift [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+        namedIDs[name] = id
+        return id
+    }
+
+
+    func renderScreenStateNode(name: String, isDismissedOnUse dismissOnUse: Bool) {
+        let id = self.id(for: name)
+        var styleCode = "label=\"\(name)\"; "
+        if dismissOnUse {
+            styleCode += "fillcolor=lightgray; color=gray; style=filled"
```
Comment (justindarc) [inline Sources/GraphRepresentation.swift]: nit: might be nice to pull out these colors into global constants, could eventually make it customizable?

File: Sources/GraphRepresentation.swift [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+        var styleCode: String = "[ "
+
+        if let label = label {
+            styleCode += "label=\"\(label)\"; style=dashed"
+        } else {
+            styleCode += "style=solid"
+        }
+
+        if isBackable {
+            styleCode += "; dir=both; arrowtail=obox; arrowhead=normal"
```
Comment (justindarc) [inline Sources/GraphRepresentation.swift]: nit: this is fine right now, but seems fragile... would be nice to have a utility function that lets you pass in a bunch of `key=value` strings that returns them concatenated together with proper `;` placements and whitespace

**Review by justindarc — APPROVED** [review]

File: Sources/GraphRepresentation.swift (hunk 1 above)
Comment (jhugman) [reply Sources/GraphRepresentation.swift]: 👍

File: Sources/GraphRepresentation.swift (hunk 3 above)
Comment (jhugman) [reply Sources/GraphRepresentation.swift]: Great idea!

Comment (jhugman) [reply Sources/GraphRepresentation.swift]: :+1:

File: Sources/GraphRepresentation.swift (hunk 2 above)
Comment (jhugman) [reply Sources/GraphRepresentation.swift]: I've pulled these out into style dictionaries which we merge at render time.

**Pull request merged**